lib/express/server.js

```diff
@@ -50,88 +53,44 @@ async function server (opts = {}) {
   const close = httpServer.close.bind(httpServer);
   httpServer.close = async () => await new B((resolve, reject) => {
     // https://github.com/nodejs/node-v0.x-archive/issues/9066#issuecomment-124210576
-    isServerClosed = true;
-    serverStateNotifier.emit('shutdown');
+    serverState.closed = true;
+    serverState.notifier.emit('shutdown');
     httpServer.on('close', resolve);
     close((err) => {
       if (err) reject(err); // eslint-disable-line curly
     });
   });
 
-  return await new B((resolve, reject) => {
-    httpServer.on('error', (err) => {
-      if (err.code === 'EADDRNOTAVAIL') {
-        log.error('Could not start REST http interface listener. ' +
-                  'Requested address is not available.');
-      } else {
-        log.error('Could not start REST http interface listener. The requested ' +
-                  'port may already be in use. Please make sure there is no ' +
-                  'other instance of this server running already.');
-      }
+  return await new B(async (resolve, reject) => {
+    // we put an async function as the promise constructor because we want some things to happen in
+    // serial (application of plugin updates, for example). But we still need to use a promise here
+    // because some elements of server start failure only happen in httpServer listeners. So the
+    // way we resolve it is to use an async function here but to wrap all the inner logic in
+    // try/catch so any errors can be passed to reject.
+    try {
+      configureHttp({httpServer, serverState, reject});
```

Comment:
these two synchronous methods could be extracted out of the promise

Reply:
I don't think so, since they need the `reject` method, right?

Reply:
ah, I see now. Makes sense